### ISSUE TITLE
Tracking Issue: upgrade to 128-bit amount values in notes

### ISSUE THREAD

**Issue by avahowell** [penumbra-zone/penumbra#1378]
Following #1336, where we have decided to move to 128 bit amounts in our notes, we need to do the implementation. A few steps are as follows

- [ ] Update the spec to specify how we handle bridged assets, where we will return an error if we can't fit an asset in a u128, and the exact encoding for the u128
- [x] Implement a new `Amount` domain type that encapsulates a u128, and provides functionality for encoding the u128 as two u64s for use in our protobufs.
- [x] Propagate the new `Amount` type throughout the codebase, replacing u64 where it was used previously.

**Comment by hdevalence** [comment]
@avahowell What's the current status of this issue?

**Comment by avahowell** [comment]
We have the new amount type now, but it is still using a u64 internally

**Comment by hdevalence** [comment]
Following #1819, what's the status of this on the client side? We at least need to be able to record 128-bit values in sqlite in the rust view server (cc @aubrika).

**Comment by conorsch** [comment]
@aubrika Could you summarize your opinions on our options here? I believe there was some relevant discussion in Discord recently that you may want to cross-link.

**Comment by aubrika** [comment]
> @aubrika Could you summarize your opinions on our options here? I believe there was some relevant discussion in Discord recently that you may want to cross-link.

I think storing the u128 as a `BLOB` in SQLite will make for easier type conversions than using two i64 columns & will still allow us to do sorting & range selects on the `BLOB` values. I don't think there's any issue yet for the needed storage changes but the needed changes shouldn't be too complicated.

**Comment by conorsch** [comment]
@erwanor Can you take another look and just comment on current status? I suspect this work will need to be tracked on the next testnet, given @aubrika's comment above, but I defer to you on that.

**Comment by redshiftzero** [comment]
One idea mentioned in sprint planning was to add another token to genesis with a high level of precision similar to eth (e18) to see what breaks. This could be done with a single note.

**Comment by erwanor** [comment]
+ We'll also need to change the signature to `Amount::value -> u128` and integrate that with the rest of the codebase

**Comment by hdevalence** [comment]
Sounds good, I think we'll generally want to avoid converting to u64s, and migrating our existing uses to use Amounts

**Comment by hdevalence** [comment]
This has been done for a while.
